File: OP_lab/lab4/c++/myString.cpp
//
// Created by maxim on 22.02.2020.
//

#include <cstring>
#include "myString.h"

myString::myString() : size{0} {
    str = new char[0];
}

myString::myString(char *str) {
    initialization(str);
}

myString::myString(const myString & other) {
    initialization(other.str);
}

myString::~myString() {
    delete [] str;
}

std::ostream &operator<<(std::ostream &out, const myString &str) {
    out << str.str;
    return out;
}

void myString::initialization(char *other) {
    size = strlen(other);
    this->str = new char[size + 1];
    for (int i = 0; i < size; ++i) {
        this->str[i] = other[i];
    }
    this->str[size] = '\0';
}

myString &myString::operator=(const myString &other) {
    this->size = other.get_size();
    if(this->str != nullptr)
        delete[] this->str;
    this->str = new char[size + 1];
    for (int i = 0; i < size + 1; ++i) {
        this->str[i] = other.str[i];
    }
    return *this;
}
std::size_t myString::get_size() const {
    return size;
}

myString myString::operator*(int n) const{
    if(size == 0) return myString(str);
    std::size_t len{size * n};
    char *tmp = new char[len];
    for (int i = 0; i < len; ++i) {
        tmp[i] = str[i % size];
    }
    return myString(tmp);
}

myString myString::operator+(const myString &other) const {
    std::size_t len = this->size + other.size;
    char *tmp = new char[len];
    for (int i = 0; i < this->size; ++i) {
        tmp[i] = this->str[i];
    }
    for (int i = this->size; i < len; ++i) {
        tmp[i] = other.str[i - this->size];
    }
    return myString(tmp);
}

File: OP_lab/lab4/c#/myString.cs
using System;
using System.IO;

namespace lab_2_OP_
{
    public class MyString
    {
        //private:
        private char[] _str;
        //public:
        public MyString()
        {
            _str = new char[0];
        }

        public MyString(char[] ch)
        {
            this._str = ch;
        }

        public MyString(MyString other)
        {
            this._str = other._str;
        }

        public int GetSize()
        {
            return _str.Length;
        }

        public override string ToString()
        {
            return new string(_str);
        }

        public static MyString operator *(MyString str1, int n)
        {
            int len = str1.GetSize() * n;
            char[] tmp = new char[len];
            for (int i = 0; i < len; i++)
            {
                tmp[i] = str1._str[i % str1.GetSize()];
            }

            return new MyString(tmp);
        }

        public static MyString operator +(MyString str1, MyString str2)
        {
            char[] tmp = new char[str1.GetSize() + str2.GetSize()];
            for (int i = 0; i < str1.GetSize(); i++)
            {
                tmp[i] = str1._str[i];
            }
            for (int i = str1.GetSize(); i < tmp.Length; i++)
            {
                tmp[i] = str2._str[i - str1.GetSize()];
            }
            return new MyString(tmp);
        }
    }
}